x/erc20/keeper/ibc_callbacks.go

```diff
@@ -82,36 +80,51 @@ func (k Keeper) OnRecvPacket(
 	)
 
 	// check if the coin is a native staking token
-	bondDenom := k.stakingKeeper.BondDenom(ctx)
-	if coin.Denom == bondDenom {
+	if coin.Denom == evmParams.EvmDenom {
 		// no-op, received coin is the staking denomination
 		return ack
 	}
 
 	pairID := k.GetTokenPairID(ctx, coin.Denom)
-	if len(pairID) == 0 {
-		// short-circuit: if the denom is not registered, conversion will fail
-		// so we can continue with the rest of the stack
-		return ack
-	}
+	pair, found := k.GetTokenPair(ctx, pairID)
+	switch {
+	// Case 1. token pair is not registered and is a single hop IBC Coin
+	case !found && ibc.IsSingleHop(coin.Denom):
+		contractAddr, err := utils.GetIBCDenomAddress(coin.Denom)
+		if err != nil {
+			return channeltypes.NewErrorAcknowledgement(err)
+		}
+
+		found := evmParams.IsPrecompileRegistered(contractAddr.String())
+		if found {
+			return ack
+		}
 
-	pair, _ := k.GetTokenPair(ctx, pairID)
-	if !pair.Enabled {
-		// no-op: continue with the rest of the stack without conversion
+		if err := k.RegisterPrecompileForCoin(ctx, coin.Denom, contractAddr); err != nil {
+			return channeltypes.NewErrorAcknowledgement(err)
+		}
 		return ack
-	}
 
-	balance := k.bankKeeper.GetBalance(ctx, recipient, coin.Denom)
+	// TODO: Will this ever happen ?
+	// Case 2. token pair is registered but does not have a precompile registered
+	// NOTE: This should never happen as the precompile is registered on the token pair creation
+	// case found && pair.IsNativeCoin() && !evmParams.IsPrecompileRegistered(pair.GetERC20Contract().String()):
 
-	switch {
+	// Case 3. native ERC20 token
 	case pair.IsNativeERC20():
-		if err := k.LegacyConvertCoinNativeERC20(ctx, pair, balance.Amount, common.BytesToAddress(recipient.Bytes()), recipient); err != nil {
-			return channeltypes.NewErrorAcknowledgement(err)
+		// ERC20 module or token pair is disabled -> return
+		if !k.IsERC20Enabled(ctx) || !pair.Enabled {
+			return ack
 		}
-	case pair.IsNativeCoin():
-		if err := k.LegacyConvertCoinNativeCoin(ctx, pair, balance.Amount, common.BytesToAddress(recipient.Bytes()), recipient); err != nil {
+
+		balance := k.bankKeeper.GetBalance(ctx, recipient, coin.Denom)
+		if err := k.LegacyConvertCoinNativeERC20(ctx, pair, balance.Amount, common.BytesToAddress(recipient.Bytes()), recipient); err != nil {
```

Comment:
This should be skipped if the balance is 0

Reply:
It shouldnt be 0 since it just received a transfer